sdk/core/core-lro/samples-dev/customPoller.ts

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * @summary demnostrates how to create a custom poller.
+ */
+
+import type { OperationState, OperationStatus, PollerLike } from "@azure/core-lro";
+import { delay } from "@azure/core-util";
+
+const DEFAULT_POLL_INTERVAL_IN_MS = 2000;
+
+type OperationConfig = { id: string };
+
+type RestorableOperationState<TResult, TState extends OperationState<TResult>> = TState & {
+  config: OperationConfig;
+};
+
+async function initOperation<T>(): Promise<RestorableOperationState<T, OperationState<T>>> {
+  // starts the operation
+  return { status: "running", config: { id: "1" } };
+}
+
+async function pollOperation<T>({
+  setDelay,
+  state,
+  options,
+}: {
+  state: RestorableOperationState<T, OperationState<T>>;
+  setDelay: (interval: number) => void;
+  options?: { abortSignal?: AbortSignal };
+}): Promise<void> {
+  if (options?.abortSignal?.aborted) {
+    throw new Error("aborted");
+  }
+  // polls the operation using id while respecting the abort signal in options
+  const { id } = state.config;
+  const response = { id, status: "succeeded", retryAfter: 2000 };
+  // update the state based on the response
+  if (response.status === "succeeded") {
+    state.status = "succeeded";
+    state.result = response as T;
+  } else if (response.status === "failed") {
+    state.status = "failed";
+    state.error = (response as any).error;
+  } else if (response.status === "canceled") {
+    state.status = "canceled";
+  } else {
+    setDelay(response.retryAfter); // updates the delay if the service suggests a different interval
+  }
+}
+
+function deserializeState<T, TState extends OperationState<T>>(
+  serializedState: string,
+): RestorableOperationState<T, TState> {
+  try {
+    return JSON.parse(serializedState).state;
+  } catch (e) {
+    throw new Error(`Unable to deserialize input state: ${serializedState}`);
+  }
+}
+
+function createPoller<T>({
+  intervalInMs,
+  restoreFrom,
+}: {
+  restoreFrom?: string;
+  intervalInMs?: number;
+}): PollerLike<OperationState<T>, T> {
+  let statePromise: Promise<OperationState<T>>;
+  let state: RestorableOperationState<T, OperationState<T>>;
+  if (restoreFrom) {
+    state = deserializeState(restoreFrom);
+    statePromise = Promise.resolve(state);
+  } else {
+    statePromise = initOperation<T>().then((s) => (state = s));
+  }
+  let resultPromise: Promise<T> | undefined;
+  const abortController = new AbortController();
+  // Progress handlers
+  type Handler = (state: OperationState<T>) => void;
```

Comment:
Should we move this type outside the function?

Reply:
I don't believe so, it represents an internal state object.